spec/std/crystal/hasher_spec.cr

```diff
@@ -0,0 +1,147 @@
+require "spec"
+require "bit_array"
+require "random/secure"
+
+struct Crystal::Hasher
+  struct SpecResult
+    def initialize(@v = 0_u64)
+    end
+
+    def differs(other)
+      @v.should_not eq(other.@v)
+      (@v ^ other.@v).popcount.should be_close(32, 24)
+    end
+  end
+
+  def self.bytes(b)
+    SpecResult.new(new.bytes(b).result)
+  end
+end
+
+# Note: this tests are probabilistic.
+# Hasher is randomly seeded, therefore this tests could randomly
+# fail. Probability of fail is very very low, but not zero.
+# If test fails two times in a row, then error is certainly real.
+describe "Hasher" do
+  describe "bytes" do
+    it "should differ for every single bit and length" do
+      empty_hash = Crystal::Hasher.bytes(Bytes.new(0))
+      1.upto(64) do |byte_len|
+        bytes = Bytes.new(byte_len, 0_u8)
+        zero_hash = Crystal::Hasher.bytes(bytes)
+        prev_hash = zero_hash
+        prev_prev_hash = Crystal::Hasher::SpecResult.new
+        bit_len = byte_len * 8
+        0.upto(bit_len - 1) do |bit|
+          bytes[bit/8] ^= 1 << (bit & 7)
+          cur_hash = Crystal::Hasher.bytes(bytes)
+          cur_hash.differs empty_hash
+          cur_hash.differs zero_hash
+          cur_hash.differs prev_hash
+          cur_hash.differs prev_prev_hash
+          prev_hash, prev_prev_hash = cur_hash, prev_hash
+          # check for other random bit
+          while true
```

Comment:
```cr
loop do
  # ...
end
```

Reply:
Only after you change documentation:
https://crystal-lang.org/docs/syntax_and_semantics/while.html
"Checking the condition at the end of a loop"

Reply:
Change... what?

Reply:
Cite from https://crystal-lang.org/docs/syntax_and_semantics/while.html :

## Checking the condition at the end of a loop

If you need to execute the body at least once and then check for a breaking condition, you can do this:

```crystal
while true
  do_something
  break if some_condition
end
```

Reply:
it is followed by:

Or use `loop`, found in the standard library:

```crystal
loop do
  do_something
  break if some_condition
end
```

But since `while true` is also mentioned, and it is mentioned first, and I prefer raw statements instead of unnecessary abstractions, I have all rights to use `while true` here.

Reply:
I guess it's all about semantics and `loop` is pretty self-explanatory, but please treat it as a suggestion, stdlib uses both variants FYI.

```bash
crystal git:master ❯ grep -r "while true" . | wc -l
# => 161
crystal git:master ❯ grep -r "loop do" . | wc -l
# => 67
```

Reply:
@Sija - `while true` is OK.

Reply:
It's OK until we standardize it. I might send a latter PR to make it all `loop do`. But lets not discuss it here.